File: go/test/wonder_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/world-wonders-sdk/go"
	"github.com/voxgig-sdk/world-wonders-sdk/go/core"

	vs "github.com/voxgig-sdk/world-wonders-sdk/go/utility/struct"
)

func TestWonderEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Wonder(nil)
		if ent == nil {
			t.Fatal("expected non-nil WonderEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := wonderBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "wonder." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set WORLDWONDERS_TEST_WONDER_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		wonderRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.wonder", setup.data)))
		var wonderRef01Data map[string]any
		if len(wonderRef01DataRaw) > 0 {
			wonderRef01Data = core.ToMapAny(wonderRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = wonderRef01Data

		// LIST
		wonderRef01Ent := client.Wonder(nil)
		wonderRef01Match := map[string]any{}

		wonderRef01ListResult, err := wonderRef01Ent.List(wonderRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, wonderRef01ListOk := wonderRef01ListResult.([]any)
		if !wonderRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", wonderRef01ListResult)
		}

		// LOAD
		wonderRef01MatchDt0 := map[string]any{
			"id": wonderRef01Data["id"],
		}
		wonderRef01DataDt0Loaded, err := wonderRef01Ent.Load(wonderRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		wonderRef01DataDt0LoadResult := core.ToMapAny(wonderRef01DataDt0Loaded)
		if wonderRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if wonderRef01DataDt0LoadResult["id"] != wonderRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func wonderBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "wonder", "WonderTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read wonder test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse wonder test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"wonder01", "wonder02", "wonder03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("WORLDWONDERS_TEST_WONDER_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"WORLDWONDERS_TEST_WONDER_ENTID": idmap,
		"WORLDWONDERS_TEST_LIVE":      "FALSE",
		"WORLDWONDERS_TEST_EXPLAIN":   "FALSE",
		"WORLDWONDERS_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["WORLDWONDERS_TEST_WONDER_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["WORLDWONDERS_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["WORLDWONDERS_APIKEY"],
			},
			extra,
		})
		client = sdk.NewWorldWondersSDK(core.ToMapAny(mergedOpts))
	}

	live := env["WORLDWONDERS_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["WORLDWONDERS_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: ts/test/entity/wonder/WonderEntity.test.ts


import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'
import { createLiveTransport } from '../../live-runner'
import { runLiveEntity } from '../../live-entity'


import { WorldWondersSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveClientOptions,
  liveDelay,
  loadEnvLocal,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


// AFTER the imports on purpose: TypeScript hoists `import` above any
// statement in the emitted CommonJS, so a loader placed above them would
// run only after every imported module had already been evaluated - and
// anything reading process.env at module scope would miss these values.
loadEnvLocal(__dirname + '/../../../.env.local')


describe('WonderEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when WORLD_WONDERS_TEST_LIVE=TRUE.
  afterEach(liveDelay('WORLD_WONDERS_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = WorldWondersSDK.test()
    const ent = testsdk.Wonder()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.WORLD_WONDERS_TEST_LIVE
    for (const op of ['list', 'load']) {
      if (!live && maybeSkipControl(t, 'entityOp', 'wonder.' + op, live)) return
    }

    
    const setup = basicSetup()
    if (setup.live) {
      return runLiveEntity(setup, {"active":true,"alias":{"field":{}},"fields":[{"active":true,"name":"build_year","req":false,"short":"Year the wonder was built","type":"`$INTEGER`","index$":0},{"active":true,"name":"id","req":false,"short":"Unique identifier for the wonder","type":"`$STRING`","index$":1},{"active":true,"name":"links","req":false,"type":"`$OBJECT`","index$":2},{"active":true,"name":"location","req":false,"type":"`$OBJECT`","index$":3},{"active":true,"name":"name","req":false,"short":"Name of the world wonder","type":"`$STRING`","index$":4},{"active":true,"name":"summary","req":false,"short":"Brief summary of the wonder","type":"`$STRING`","index$":5},{"active":true,"name":"time_period","req":false,"short":"Historical time period of the wonder","type":"`$STRING`","index$":6}],"id":{"field":"id","name":"id"},"name":"wonder","op":{"list":{"input":"data","name":"list","points":[{"active":true,"args":{"query":[{"active":true,"example":10,"kind":"query","name":"limit","orig":"limit","reqd":false,"type":"`$INTEGER`","index$":0},{"active":true,"example":0,"kind":"query","name":"offset","orig":"offset","reqd":false,"type":"`$INTEGER`","index$":1}]},"contract":{"id":"GET /wonders","json":"{\"operationId\":\"getAllWonders\",\"parameters\":[{\"description\":\"Maximum number of wonders to return\",\"in\":\"query\",\"name\":\"limit\",\"required\":false,\"schema\":{\"default\":10,\"maximum\":100,\"minimum\":1,\"type\":\"integer\"}},{\"description\":\"Number of wonders to skip\",\"in\":\"query\",\"name\":\"offset\",\"required\":false,\"schema\":{\"default\":0,\"minimum\":0,\"type\":\"integer\"}}],\"protocol\":\"http\",\"responses\":{\"200\":{\"content\":{\"application/json\":{\"schema\":{\"items\":{\"properties\":{\"build_year\":{\"description\":\"Year the wonder was built\",\"type\":\"integer\"},\"id\":{\"description\":\"Unique identifier for the wonder\",\"type\":\"string\"},\"links\":{\"properties\":{\"britannica\":{\"description\":\"Britannica link\",\"format\":\"uri\",\"type\":\"string\"},\"google_maps\":{\"description\":\"Google Maps link\",\"format\":\"uri\",\"type\":\"string\"},\"trip_advisor\":{\"description\":\"TripAdvisor link\",\"format\":\"uri\",\"type\":\"string\"},\"wiki\":{\"description\":\"Wikipedia link\",\"format\":\"uri\",\"type\":\"string\"}},\"type\":\"object\"},\"location\":{\"properties\":{\"city\":{\"description\":\"City where the wonder is located\",\"type\":\"string\"},\"continent\":{\"description\":\"Continent where the wonder is located\",\"type\":\"string\"},\"coordinates\":{\"properties\":{\"latitude\":{\"format\":\"double\",\"type\":\"number\"},\"longitude\":{\"format\":\"double\",\"type\":\"number\"}},\"type\":\"object\"},\"country\":{\"description\":\"Country where the wonder is located\",\"type\":\"string\"}},\"type\":\"object\"},\"name\":{\"description\":\"Name of the world wonder\",\"type\":\"string\"},\"summary\":{\"description\":\"Brief summary of the wonder\",\"type\":\"string\"},\"time_period\":{\"description\":\"Historical time period of the wonder\",\"type\":\"string\"}},\"type\":\"object\"},\"type\":\"array\"}}},\"description\":\"Successful response\"},\"400\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"error\":{\"description\":\"Error message\",\"type\":\"string\"},\"status\":{\"description\":\"HTTP status code\",\"type\":\"integer\"}},\"type\":\"object\"}}},\"description\":\"Bad request\"},\"500\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"error\":{\"description\":\"Error message\",\"type\":\"string\"},\"status\":{\"description\":\"HTTP status code\",\"type\":\"integer\"}},\"type\":\"object\"}}},\"description\":\"Internal server error\"}},\"securitySource\":\"unspecified\"}","source":"openapi3","version":1},"kind":"http","method":"GET","orig":"/wonders","segments":[{"lit":"wonders"}],"select":{"exist":["limit","offset"]},"transform":{"req":"`reqdata`","res":"`body`"},"index$":0}],"key$":"list"},"load":{"input":"data","name":"load","points":[{"active":true,"args":{"params":[{"active":true,"kind":"param","name":"id","orig":"id","reqd":true,"type":"`$STRING`","index$":0}]},"contract":{"id":"GET /wonders/{id}","json":"{\"operationId\":\"getWonderById\",\"parameters\":[{\"description\":\"Unique identifier of the world wonder\",\"in\":\"path\",\"name\":\"id\",\"required\":true,\"schema\":{\"type\":\"string\"}}],\"protocol\":\"http\",\"responses\":{\"200\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"build_year\":{\"description\":\"Year the wonder was built\",\"type\":\"integer\"},\"id\":{\"description\":\"Unique identifier for the wonder\",\"type\":\"string\"},\"links\":{\"properties\":{\"britannica\":{\"description\":\"Britannica link\",\"format\":\"uri\",\"type\":\"string\"},\"google_maps\":{\"description\":\"Google Maps link\",\"format\":\"uri\",\"type\":\"string\"},\"trip_advisor\":{\"description\":\"TripAdvisor link\",\"format\":\"uri\",\"type\":\"string\"},\"wiki\":{\"description\":\"Wikipedia link\",\"format\":\"uri\",\"type\":\"string\"}},\"type\":\"object\"},\"location\":{\"properties\":{\"city\":{\"description\":\"City where the wonder is located\",\"type\":\"string\"},\"continent\":{\"description\":\"Continent where the wonder is located\",\"type\":\"string\"},\"coordinates\":{\"properties\":{\"latitude\":{\"format\":\"double\",\"type\":\"number\"},\"longitude\":{\"format\":\"double\",\"type\":\"number\"}},\"type\":\"object\"},\"country\":{\"description\":\"Country where the wonder is located\",\"type\":\"string\"}},\"type\":\"object\"},\"name\":{\"description\":\"Name of the world wonder\",\"type\":\"string\"},\"summary\":{\"description\":\"Brief summary of the wonder\",\"type\":\"string\"},\"time_period\":{\"description\":\"Historical time period of the wonder\",\"type\":\"string\"}},\"type\":\"object\"}}},\"description\":\"Successful response\"},\"404\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"error\":{\"description\":\"Error message\",\"type\":\"string\"},\"status\":{\"description\":\"HTTP status code\",\"type\":\"integer\"}},\"type\":\"object\"}}},\"description\":\"Wonder not found\"},\"500\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"error\":{\"description\":\"Error message\",\"type\":\"string\"},\"status\":{\"description\":\"HTTP status code\",\"type\":\"integer\"}},\"type\":\"object\"}}},\"description\":\"Internal server error\"}},\"securitySource\":\"unspecified\"}","source":"openapi3","version":1},"kind":"http","method":"GET","orig":"/wonders/{id}","segments":[{"lit":"wonders"},{"var":"id"}],"select":{"exist":["id"]},"transform":{"req":"`reqdata`","res":"`body`"},"index$":0}],"key$":"load"}},"relations":{"ancestors":[]},"key$":"wonder","name__orig":"wonder","Name":"Wonder","name_":"wonder","name-":"wonder","NAME":"WONDER","index$":0}, {"active":true,"entity":"wonder","key$":"BasicWonderFlow","kind":"basic","name":"BasicWonderFlow","param":{},"step":[{"active":true,"data":{},"input":{},"match":{},"op":"list","spec":[],"valid":[{"apply":"ItemExists","def":{"ref":"wonder_ref01"}}],"index$":0},{"active":true,"data":{},"input":{"ref":"wonder_ref01","srcdatavar":"wonder_ref01_data","suffix":"_dt0"},"match":{"id":"wonder01"},"op":"load","spec":[],"valid":[{"apply":"TextFieldMark","def":{"mark":"Mark01-wonder_ref01"}}],"index$":1}]}, 'Wonder')
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let wonder_ref01_data = Object.values(setup.data.existing.wonder)[0] as any

    // LIST
    const wonder_ref01_ent = client.Wonder()
    const wonder_ref01_match: any = {}

    const wonder_ref01_list = (await wonder_ref01_ent.list(wonder_ref01_match)).map((e: any) => e.data())


    // LOAD
    const wonder_ref01_match_dt0: any = {}
    wonder_ref01_match_dt0.id = wonder_ref01_data.id
    const wonder_ref01_data_dt0 = (await wonder_ref01_ent.load(wonder_ref01_match_dt0)).data()
    assert(wonder_ref01_data_dt0.id === wonder_ref01_data.id)


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/wonder/WonderTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = WorldWondersSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['wonder01','wonder02','wonder03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'WORLD_WONDERS_TEST_WONDER_ENTID': idmap,
    'WORLD_WONDERS_TEST_LIVE': 'FALSE',
    'WORLD_WONDERS_TEST_EXPLAIN': 'FALSE',
  })

  idmap = env['WORLD_WONDERS_TEST_WONDER_ENTID']

  const live = 'TRUE' === env.WORLD_WONDERS_TEST_LIVE

  const transport = createLiveTransport()
  if (live) {
    const rawIds = process.env['WORLD_WONDERS_TEST_WONDER_ENTID']
    idmap = rawIds && rawIds.trim() ? JSON.parse(rawIds) : {}
    if (!idmap || Array.isArray(idmap) || typeof idmap !== 'object') {
      throw new Error('Live ENTID must be a JSON object')
    }
    client = new WorldWondersSDK(merge([
      // FIRST, so the generated fields below win: sdk-test-control.json's
      // test.client.options adds to the live client, it does not redirect it.
      liveClientOptions(),
      {
      },
      // 'extra || {}', not a bare 'extra': struct.merge returns UNDEFINED when the
      // last entry is undefined, and basicSetup is normally called with no
      // argument at all - so a bare 'extra' silently discarded the apikey
      // and server values above and handed the SDK undefined. Harmless
      // while there was nothing in that object; not harmless now.
      extra || {},
      { system: { fetch: transport.fetch } }
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.WORLD_WONDERS_TEST_EXPLAIN,
    live,
    transport,
    now: Date.now(),
  }

  return setup
}
  
